hashing/sha256.cpp

```diff
@@ -0,0 +1,227 @@
+    uint64_t input_size = input.length() * 8;  // Message length in bits
+
+    // Append a single '1' bit
+    padded_input += '\x80';
+
+    // Pad with zeros
+    while ((padded_input.length() * 8) % 512 != 448) padded_input += '\x00';
+
+    // Append length of the original input string as a 64-bit big-endian integer
+    for (uint32_t i = 64; i != 0; i -= 8) {
+        padded_input += static_cast<char>((input_size >> (i - 8)) & 0xFF);
+    }
+
+    // Process message in successive 512-bit (64-byte) chunks
+    for (size_t i = 0; i < padded_input.length(); i += 64) {
+        uint32_t blocks[64];
+
+        // Copy chunk into first 16 words of the message schedule array
+        for (size_t j = 0; j < 16; ++j) {
+            blocks[j] =
+                (static_cast<uint8_t>(padded_input[i + j * 4]) << 24) |
+                (static_cast<uint8_t>(padded_input[i + j * 4 + 1]) << 16) |
+                (static_cast<uint8_t>(padded_input[i + j * 4 + 2]) << 8) |
+                static_cast<uint8_t>(padded_input[i + j * 4 + 3]);
+        }
+
+        for (size_t j = 16; j < 64; ++j) {
+            uint32_t s0 = rightRotate(blocks[j - 15], 7) ^
+                          rightRotate(blocks[j - 15], 18) ^
+                          (blocks[j - 15] >> 3);
+            uint32_t s1 = rightRotate(blocks[j - 2], 17) ^
+                          rightRotate(blocks[j - 2], 19) ^
+                          (blocks[j - 2] >> 10);
+            blocks[j] = blocks[j - 16] + s0 + blocks[j - 7] + s1;
+        }
+
+        // Initialize working variables
+        uint32_t a = hash[0];
+        uint32_t b = hash[1];
+        uint32_t c = hash[2];
+        uint32_t d = hash[3];
+        uint32_t e = hash[4];
+        uint32_t f = hash[5];
+        uint32_t g = hash[6];
+        uint32_t h = hash[7];
+
+        // Compression function main loop
+        for (size_t j = 0; j < 64; ++j) {
+            uint32_t s1 =
+                rightRotate(e, 6) ^ rightRotate(e, 11) ^ rightRotate(e, 25);
+            uint32_t ch = (e & f) ^ (~e & g);
+            uint32_t temp1 = h + s1 + ch + k[j] + blocks[j];
+            uint32_t s0 =
+                rightRotate(a, 2) ^ rightRotate(a, 13) ^ rightRotate(a, 22);
+            uint32_t maj = (a & b) ^ (a & c) ^ (b & c);
+            uint32_t temp2 = s0 + maj;
+
+            h = g;
+            g = f;
+            f = e;
+            e = d + temp1;
+            d = c;
+            c = b;
+            b = a;
+            a = temp1 + temp2;
+        }
+
+        // Update hash values
+        hash[0] += a;
+        hash[1] += b;
+        hash[2] += c;
+        hash[3] += d;
+        hash[4] += e;
+        hash[5] += f;
+        hash[6] += g;
+        hash[7] += h;
+    }
+
+    // Convert the hash to a hexadecimal string
+    std::string result;
+    for (size_t i = 0; i < 8; ++i) {
+        for (int j = 0; j < 4; ++j) {
+            uint8_t byte = (hash[i] >> (24 - j * 8)) & 0xFF;
+            char buf[2];
```

Comment:
```suggestion
            char buf[3];
```
`buf` is one `char` [too short](https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMQByARg9KtQYEAysib0QXACx8BBAKoBnTAAUAHpwAMvAFYTStJg1DIApACYAQuYukl9ZATwDKjdAGFUtAK4sGIMwDMpK4AMngMmAByPgBGmMQSAKxcpAAOqAqETgwe3r7%2BQemZjgJhEdEscQlcybaY9iUMQgRMxAS5Pn6BdQ3Zza0EZVGx8UkpCi1tHfndEwNDFVVjAJS2qF7EyOwcAPQAVHsmGgCCANR7pwACNPRHZxeXTF4ECCSnJokWALLoAHSnxwYeAUXlopwAEkwAI5eTAfAAiEAQBAIqQUIB2O2AhAQXhivzQLB2LHQhmBoIQyzu5yuMWIeEwVFOQjwLFS9FObms1lOrPZmDYgiYjVOqCZL0w70%2BQnBxwAtGZEgA2CFMBQIcLAAG0YAkHEsBHUi5IlFojE7Ri/ADueAA1nhUph8ExfiRgDsbfadjL5WYqSdzkarlgWnQFEGPhYfQqESbUejMZbPQ6nXgXW6PXa8N7ZQrlryFKcmKclARRUzkMQAJ6pIjAYhMVIa5CnBBqhBBqheBgObKFl7C05WtWnLCZYARdCnGJV04SiOfSJCY6x5Hx81JrOO52u4ju5M7SLC7JiAD6QkwyA2hCrp%2BOwEYyCr%2BcMU5oxAmQdSeNowIQTt5BhTjMDQNC4f5o0VFVgSLU5UgGcs53/ZlczMU4qFYOgqwg1DlQLIc8CwWgqyDLwlDfN4nglQQ8FERwjBLMUCGHYhJXg5BbSYB9C1fEtLw2Ni1QUK0SCnNt1U1X4jR2O5ZMCcJkG8LB3gCNxRAUJQ2hMAJsFOPTMR2U4ADFKI0%2BICDkgIFKUyVtLUiZ8EEbTdP0zFjLeLxwgIAAOU8CFIU5PMEAIzD8osGCnIKCCVSQwtJFo5xrTBwxOcwrJ7GyVLcJwJlY1hnNcwyTOIU4AEkAHlRUdBtGhSlcTn2Q4AweZg2AUdjJXEjUjCDS46QZJlIQkhixF1ekXhYOrzhkk5WuSjrW3bTV3gAdisBqDl6ub2qYLYUN9ZVev6xljO7XsBELfgSolKUo1wpVV1NBMLQYa0t1TdM90zL1IP9e4ix1PUJt5Nl6EFFpGmku5toW9UmCg1b1uORqjvpE6ACVUAh5KkMlGJCELMVYJCuV8bLbtx0nQCCEwB9iF6%2BCGxYU4gNKwRafiOdUFOYgseFOFmquRnWB5vmadORZOaJsn%2B253nsd61iCA2ICopCsKABVkPl/mpy8jn6eambjjV0Ky3pYBkUx7GIFNsKGACzIAC9MDCnWafzEw1upPSecwZXiCAiAgOcgr3cwT2VrcU5g6yuyY5C045VF7HlipAIka9%2BFZI2pr/r6tGmS1yVILw0agYQA1BcuYXmfCb8y2LwCG5LAh6QYohFvVRX/ZV1v0BAEBcuWpuaGYMEutOAA3MRYShk4HMH4eGLhqCIEXoe2%2BW8wlXr55I6R33TjctmsjEPAXaLYgG1nInJ5n7wcZtF50LwD8y0TmXEJuqgG3O8efauTggMQm4pkIKBhK0SUvMsagNxq/d%2BpxvJwXpG1YCvxfhcAAJyAMCl5dWZYuqRm8giFS8JEYaFUEqY4GgsHYCVEqFaAVKEWAsIw442BvKJGYaoVSSpsBGQCCtMwPDjiJEkEZIyiQAjHFILgo%2BCjFEKMockDQ2BEhmBWkZHhWCLAaGVN5bybgeFcCMt5AI8IsHHBsKcFRFhsAaDcPCbBWdtKZwDL7E%2BQJHDn0vq0G%2BiFebdinGgBgcxBCFmfggBBExTifwJt/ZCv9dqNDELgtyjMCBwJuleOIotYGJMlO%2BWJMUUGshxmYDBAQuBcFwaE2Jdsyy2kjDFUh2lyFe0PkoyhkgzDeWOGYIyWDvI8JWlwIRkhJBYJSLY1QFhEhuA0EZCwbhtGzOwLoxI8ILBiJ4QELByo3CKhsSorBRkammJSPI32lCsGBDMWYY40hZnWK4G4RI2B4TcNmfCbyGgVrHGOMMkxfSAifDAjwswkhqlcPsTwxIiRHErXhG4II1y9KUOEfYrZK1nmUL%2BfCbAFhTHYB0RYFFGhqFMNmW4bBFhzm4p4dgKZbCsG0qZcs5lK1vJKjkR4xRlClm0qwSi3lsyoWOOOG84x4rCV3LcEqNZPTji4u8pIQF8K3B6KBSi0ZirhneXhLI9Fszhk6WSBo0R5i3kMPhDwvRGgQpRxGbM%2BliQsFaLcNSyhCqHEaHpUEH5iQVXTMkEwk1gqlRuGOEqMFMzKFSDMHcrBjDIVrRWicLhkLsBcAsGYcZLqenwjMAq%2BERkZUqICAEAl4y%2BVnG6VQxFKqwV4tUCtBhNDrEnNUN5N5Ja7mktNZoswJas0vMsEZThUqrVcBjTFGwEbVBHMkBYbyFgM08K9VG5IxxA2UOcQO3tWCeHwiVFgrBlKoU8KMpINRYKx0JspYC/5Gg61KNmdgp5tKuBioTdgIRUaNCFtUJoyQ3kVq4orbwld/qtXfMofst5jiLB7tUJIT5/SnnGv5co1Qnw2XRokTwpU3kzD0Mkah1VDzsCDsxf05UARFU8LVW4cDvbW1GK9RoECIzF0XvsZIoysiXmSERVGol9qBFWICEZb1S7GFcG5UZMwriM453rUA5QrE5SpF5lsDSmpcEb2XlqeC6AsDoFPHvMs7Tm7PDcbgqKMUwrWdPM7WyFi7MEF%2BPQIwLwID5guCQjOQCvjJQUFxSUvngAv3CNOAm6mj5uWOKkR0EVYKZCMByMA3AcvxYsjhszFmrMMBbtyTzOWPhuB7RoSranUo4bcsoJgU4omnBdrzKavsrQag5FAIrToSsNx84wGLCAAs0m8p7RUpxkhoTAGAWzkypvAPM4N6z7xLC2cqws1QoE6vuI08fQyKW0tTmiy/ImN09TYnHl51u7ctQjmLDFUmhB4vADlK4NMQF9Z01wVdGOjTeRkNOK0kLeBTgLdsxoNxIO5S2ZW50k1A3LMbfK%2BQuYjhkCnnUhZVSyA2z0x0lAVz7mVLYAKhASHycpvTZVJQyR6cul6VcQ1o7TW9PhdOG1CLD5AIli8MgfTmQp6Sjm29ssEBXszg9qcQn3ZbRdb0oD9eF9XZlkhzDuHWvVKreK9ZkbfnxvM5BxjsHkgD4muBzEWgqAOLhk%2BK0rZDn2fvrch4VIs4FcMFtNTbmxSyw/qHKJbJyFeeRZLITp0oJJT%2BKYCRHDR9VfubCtoUHsOQvp/jj%2BuH3JLDaCt0npRtv7dK8jNoNpFiTXvvXhDWiuO1T47cFFXy%2BPsAQFR0N%2BznwteWFOOni4kgESe1UvHKFkcuTF/fXpOvx4cd47sq3vyzlO8tYN6VnvFg%2B8WAHzSSQm3d%2B1K2aPrkeuf2T5r0ouf2PG%2BflUsv9va%2B1to83xZXvh%2B99D8/ypk/cc9dI5RxX6KJY4N6L4P5eRt6r5d7WaRg75f6nAH48gBAj6u5HZHxs71ToG%2Bwp7q5p6g655Z5ZQW555bYF5F7YGKLA4KAaCg4WxWxiyYAQCl4O4V5JynA1AIgBQrSewfLAEz6KL0EEDWz8zMF26sGfDp7JycFbIBRcBI58HT4CEz5iFl6O4WBSEcGJBV6U46RxLM78HUFcB0F4CWzCGMGqESEaHsG/52ocE8FSiU5KHKFKJCEiE0yWHl6SE2FcEcFYK8FOGUEuHX4sFeHWHJy2EU4FRcAaAGHOG%2ByhHqGV5bKg6JFsHSEPQpE8g0Gf5pHeHJwrRtL94KC1L1ZBGYEmpeJny/iXwiTED2gMQzz0hMC27JTW74JmywS2ZELgqGhlHvo26g49HEp9Es4KLA4tjdHtiRi2FoEDEdFxRDHTGfAoEu79FKLA4ebkLDHD5rFjFHzA5MhTHdyfDaF7HtHBSdFajHEdhO6jEXEEAEKthLEnHrr3Fu5KIe6oBsisQGYCDoRnQigsBMBxZ26oCpAmq4Euz4Ha5EHxzg67754aEUEuFGFkL8FHxuEWGYABRKgBE8ymEME2w4kcFcD4lYnEkBSKhxFBEHELFliE6g4QC2RmAqhUD4kQAAB%2BLJKowANJqJ9Jc4AoqQxh3Rn%2BJRn%2BjJPIzSkhRRu%2BeRGh7xtJvs1BtB7SGJvsFJohTAVJ5JhJ5hNsOpHBAQepZh7hTBRpI6/JyhwOwJ2enmEAxYO804HJTprJ8uHJMQm2KoyA1pAhwONMbIaEtmORPIdpaB/B0StmwAcxyh1xnmVAsZAhRxnmcI6xAhWxo4n%2BgZIpSZM%2BU4tmpg6ZM%2BkxnmMQeZ76XptmTAFZSiTpnmOZop/eOZKmxZGByKiWnxhkcgqQ8UnU7Y08s8bRShwxsOWRW2nmNZbZvswxx%2BHSE55C5Z05ekwxkR5uRZ%2BxM5yxKGcptm6AtZK525ux85VgqZB5Xctx8yu5CZ55wxmRJ50Zt525hR45p52xeZFRjWhkHgDAYubQ8Ck8ncxY/4qgLWl4rIYgD2hmOGxmW8DEvxoI%2BOYxUJGuZunmmeu%2Buu0cwWiJZB2%2BKJSiquXke%2BsJu%2B9p0cw%2BIWSJheiMhhkBYUsumZSI25eAOhVOUK7Bg%2BSCac3psyki55R8hOrQ04XgiZnwsxy5Ci7U7cBAVAzBolAU5gKmio3GqgSlAUjFfpM%2BCFtANmC5IliZklmB7Z2cHxekSsfcOlSFskHZqUyKQCMMu0koq8h0dl5Cx2LMrA80TlF50FWBKM1cx0TIF4tAVAcoNMsSfIYMjA9eF0PcAcYS08qAhE88xwoBLYU8yVU4EVBAE2yOX5HBsSGsyUBWR2sFj292tmSl1VxZ5Vy0mAqgjoDgAE6pShSlmAAQMQGgyAkySa3kVAyAXAg1UyTAVAMQVAkgyA3k56SoY1dyvSK0TAmAkgXAy1MUWCMQWCAQk1TAUy7q0yMQ3KGiMQXC2hI6eZdVDE6wqIzwLx3UoAQ8bYUES8T1yo1Ob%2BWlG8aAt1dk8cSlbMDcIAm2wZY%2Beu6OoN0cG8rgtAF1BAA8KA11/%2B0cSl2AjVl4NMA8wNSNpwDVTVGN2NUNEUMNtVcNg831NmENwN5UzwgNWNv1eu11ZWlNhN6AxNYxQk5kEAjNP11eqZaNzV6An1pNCNPNZ%2ByNI6XARV2AQgGspwygxwQgQgnyVWDAKtNVh2SWhkZg6WxVn4OG4N85ZgOV6tuCuN6NLV1ebVI6WCVAPK6AQG8haqyAK06ASo7q8MVAmATATA3VCKLWYEiQTAnVE1VAXAMQZgXVJ1I6yA6ANQrtg1iQXVVAoEPtJC51xZ3NFN2xS0RgL18Myoeda81mQt8N5N2N/1b%2BQNSl2N4NYt/cg80NsNpdiN9N4tZG/NGNVdI62NZtAtBNwtjdJNzdotf1I61NN1BAXdINddmd/d8Ng97NZkbQXNNNPN2cfNeNToJdZNLdlNSl2tU4Gs0tst8titytCyqtF9JtZlHlAQxAh9JVuCBtwN4I9QdupwAA6iQLQPuenWMb3fja1eUSOkwO6lwEwEqOgLeuNb0hoLepID7TUitFWsgGNStDENghoK7WYMgG7eNRoDEDHUHRHYkOg5oi7Vgi1jbVIEqHCH/bgrPTcZqIXQXY9fnUqO9Q3NvSLRTXXRXbTdXa3ZVczQPUTU3TvSPXrijR3U6FPT3TIwWSI/PWI0PRI7w6PWYOPQI93UI4w0ow3So4vZpLlXo%2BveQgA1veIzw%2BXSOnfYfcfXLQrUrdnFfVffQzfW5JIC/LraVUfM/Upa/bQO/V/cQD/dDu40dhY4o6ZcA5orbZ1T%2BqQ1QKHQNdIptWYOgINext5EwFwK7YkOgANRk5IBgxhAEK7RBn0gg%2BZugCBGBKyegBemetfUdqY75bnWw89Z029cXVY2XUI/w88HI0I7XfHCzWzUZsLf03vSOqjZvZjYI5TVE3PQY6zX07vXw2PavZPXTZTXo3XeM3mRzcvaY7ZlE9w9M5s1Ct4w46fc42rW462ZnLZf5ZtIFYXKcF8CCUBF2D2JDNXBZYHIWLQf8Q1YQKlcRcCeEHld7DhjlQFnDh5ejN2HxKFeFSVSDPyODMeHFR49%2BWREQMzADfZjBcLSZl5us5I23dgOzCVMWBHvzp3Hkl1LI8DXmQ%2BAQL%2BBEHXqXeEAFL06o5vBVZPEwx0y5UqCwxwwK2MV9Rsxo6XCqENEgLs3XSK/oyAAvZ2eZb3IHLYhGcihwKsLQJwIkLwH4BwFoKQKgJwFyHhYxBsHtGlDwKQJPRa4a6sLaCAL0pgitb6xmtxiRlWtwsaxwJIGa5oLwNaxwLwOiK%2Bq61oKsHALAEgISKkHQPEOQJQKm%2BmwkMAOMikFgFPLRJgAAGoMhWjlRpacDOs3A0wfiUAxARukD4zMDVjVu8AtutBVjlQxDaDo3tukCEjgzlQMDERNtYDAlZbbCWv4CsS9hi7ohuvBCqD8Q0xNv6whuWu/h0hdseBYBNtbwsDturBUAGDAAKBluYAVtVvcC8D8CCAiBiDsBSAyCCCKAqDqBLu6ApAGC52mB4X6B4AxDoiQCrDgm1ScByhUARbeLq7tLr6/EKCRt/n0hjjwCrB2Do3ZAuARTTB%2BApChDhDDCVCjApBFBZACB4d6DkeNALAjDVA9BYcCD9BTCeCdB6CYf/wseDBEeSwMdzCsd5D4e2CTA8flD0cSAYfrCbDPtGsmvhtLtRunA9pKhyilLADC7GmYIxy4CEBvBOvLC8DxvuukCeveS/DVLTLcoHJCK4pu1BAhthukDmuWtRsxsgBxsRuJswCIAgCZ2ZsQDZv0DEBHhsCcAqdqcH4actjjKYK8CphECod6D3vCCiDiAvspfvtqBNvfukBWgNipDHv6DyfOdNtRtaO3VEwRfqeaexfGEQAeBsg5ubbVKGcutecetetYK/DSJYK3qxqiYHIZo/ucBOcueRucDueedutpzFccBmAKeueTftczerB/mZDOCSBAA%3D%3D%3D).